pkg/etcdutil/etcdutil.go

```diff
@@ -292,5 +308,98 @@ func TakeAndSaveFullSnapshot(ctx context.Context, client client.MaintenanceClose
 	metrics.SnapshotDurationSeconds.With(prometheus.Labels{metrics.LabelKind: brtypes.SnapshotKindFull, metrics.LabelSucceeded: metrics.ValueSucceededTrue}).Observe(timeTaken.Seconds())
 	logger.Infof("Total time to save full snapshot: %f seconds.", timeTaken.Seconds())
 
+	defer func() {
+		if err := os.Remove(snapshotTempDBPath); err != nil {
+			logger.Warnf("failed to remove temporary full snapshot file: %v", err)
+		}
+		if err := snapshotData.Close(); err != nil {
+			logger.Warnf("failed to close snapshot data file: %v", err)
+		}
+	}()
 	return snapshot, nil
 }
+
+// checkFullSnapshotIntegrity verifies the integrity of the full snapshot by comparing
+// the appended SHA256 hash of the full snapshot with the calculated SHA256 hash of the full snapshot data.
+func checkFullSnapshotIntegrity(snapshotData io.ReadCloser, snapTempDBFilePath string, logger *logrus.Entry) (io.ReadCloser, error) {
+	logger.Info("checking the full snapshot integrity with the help of SHA256")
+
+	// If previous temp db file already exist then remove it.
+	if err := os.Remove(snapTempDBFilePath); err != nil && !os.IsNotExist(err) {
+		return nil, err
+	}
+
+	db, err := os.OpenFile(snapTempDBFilePath, os.O_RDWR|os.O_CREATE, 0600)
+	if err != nil {
+		return nil, err
+	}
+
+	if _, err := io.Copy(db, snapshotData); err != nil {
+		return nil, err
+	}
+
+	off, err := db.Seek(0, io.SeekEnd)
+	if err != nil {
+		return nil, err
+	}
+	// 512 is chosen because it's a minimum disk sector size in most systems.
+	hasHash := (off % 512) == sha256.Size
+	if !hasHash {
+		return nil, fmt.Errorf("SHA256 hash seems to be missing from snapshot data")
+	}
+
+	var snapshotLastOffset int64
+	if snapshotLastOffset, err = db.Seek(-sha256.Size, io.SeekEnd); err != nil {
+		return nil, err
+	}
+
+	// get snapshot SHA256 hash
+	sha := make([]byte, sha256.Size)
+	if _, err := db.Read(sha); err != nil {
+		return nil, fmt.Errorf("failed to read SHA256 from snapshot data %v", err)
+	}
+
+	buf := make([]byte, hashBufferSize)
+	hash := sha256.New()
+
+	// reset the file pointer back to starting
+	currentOffset, err := db.Seek(0, io.SeekStart)
+	if err != nil {
+		return nil, err
+	}
+
+	for currentOffset+hashBufferSize < snapshotLastOffset {
+		offset, err := db.Read(buf)
+		if err != nil {
+			logger.Errorf("unable to read snapshot data into buffer to calculate SHA256: %v", err)
+			break
+		}
+
+		hash.Write(buf[:offset])
+		if currentOffset, err = db.Seek(0, io.SeekCurrent); err != nil {
+			logger.Warnf("unable to get currentOffset using seek: %v", err)
+			currentOffset += int64(offset)
+		}
+	}
+
+	if currentOffset < snapshotLastOffset {
+		if _, err := db.Read(buf); err != nil {
+			logger.Errorf("unable to read snapshot data into buffer to calculate SHA256: %v", err)
+		}
+
+		hash.Write(buf[:snapshotLastOffset-currentOffset])
+	}
```

Comment:
If an error does occur in line 386 while reading the file `db` to the buffer `buf`, we're logging an error has occurred (not a warning) - but continuing to write to the hash.

The value `buf` will not have the correct contents for us to do this.

It would be correct to return from here with an error along with an error log. Not continue.

Another suggestion, instead of calculating the number of bytes written into the buffer with `snapshotLastOffset-currentOffset`, we can not ignore the value returned by `db.Read(buf)` - which is exactly the value we want, i.e. the number of byte read. This would be easier to read.
```suggestion
	if currentOffset < snapshotLastOffset {
		if bytesRead, err := db.Read(buf); err != nil {
			logger.Errorf("unable to read snapshot data into buffer to calculate SHA256: %v", err)
			return nil, err // expand on this error
		} else {
		    hash.Write(buf[:bytesRead])
		}
	}
```